In [1]:
import pandas as pd
import numpy as np

df = pd.read_excel("Dataset for Data Analytics.xlsx")

pd.set_option('display.max_columns', None)
df.head()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


In [7]:
print("Rows and Columns:", df.shape)
print("\ncolumns")
print(df.columns)

print("\nData info:")
print(df.info())

Rows and Columns: (1200, 14)

columns
Index(['OrderID', 'Date', 'CustomerID', 'Product', 'Quantity', 'UnitPrice',
       'ShippingAddress', 'PaymentMethod', 'OrderStatus', 'TrackingNumber',
       'ItemsInCart', 'CouponCode', 'ReferralSource', 'TotalPrice'],
      dtype='object')

Data info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   object        
 1   Date             1200 non-null   datetime64[ns]
 2   CustomerID       1200 non-null   object        
 3   Product          1200 non-null   object        
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   object        
 7   PaymentMethod    1200 non-null   object        
 8   OrderStatus      1200 non-null   object        
 9   TrackingNumb

In [10]:
df.describe()

,Date,Quantity,UnitPrice,ItemsInCart,TotalPrice
count,1200,1200.000000,1200.000000,1200.000000,1200.000000
mean,2024-03-22 16:58:48,2.945833,356.412750,5.485000,1053.968300
min,2023-01-01 00:00:00,1.000000,11.390000,1.000000,11.390000
25%,2023-08-03 18:00:00,2.000000,186.062500,4.000000,410.520000
50%,2024-03-23 00:00:00,3.000000,364.210000,5.000000,823.615000
75%,2024-11-08 12:00:00,4.000000,521.570000,7.000000,1578.475000
max,2025-06-30 00:00:00,5.000000,699.930000,10.000000,3456.400000
std,NaN,1.407557,197.177146,2.281983,819.856558


In [15]:
df.isnull().sum()
print("\nMissing values in each column:")
print(missing_values)


Missing values in each column:
OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64


In [16]:
print("Full duplicate rows:", df.duplicated().sum())
print("Duplicate OrderID:", df['OrderID'].duplicated().sum())
print("Duplicate TrackingNumber:", df['TrackingNumber'].duplicated().sum())

Full duplicate rows: 0
Duplicate OrderID: 0
Duplicate TrackingNumber: 0


In [18]:
clean_df = df.copy()

change_log = []

def add_log(change_id, description, impact, status="Resolved"):
    change_log.append({
        "Change ID": change_id,
        "Description": description,
        "Impact": impact,
        "Status": status
    })

In [20]:
clean_df.columns = clean_df.columns.str.strip()

add_log(
    "CR001",
    "Column names cleaned by removing extra spaces",
    f"{len(clean_df.columns)} columns checked and standardized"
)

In [22]:
before_rows = len(clean_df)

clean_df = clean_df.drop_duplicates()

removed_duplicates = before_rows - len(clean_df)

add_log(
    "CR002",
    "Full duplicate rows removed",
    f"{removed_duplicates} duplicate rows removed"
)

print("Removed full duplicate rows:", removed_duplicates)

Removed full duplicate rows: 0


In [23]:
missing_coupon = clean_df['CouponCode'].isnull().sum()

clean_df['CouponCode'] = clean_df['CouponCode'].fillna('NO_COUPON')

add_log(
    "CR004",
    "Missing CouponCode values filled",
    f"{missing_coupon} missing CouponCode values replaced with NO_COUPON"
)

print("Missing CouponCode fixed:", missing_coupon)

Missing CouponCode fixed: 309


In [24]:
clean_df['Date'] = pd.to_datetime(clean_df['Date'], errors='coerce')

invalid_dates = clean_df['Date'].isnull().sum()

print("Invalid dates:", invalid_dates)

Invalid dates: 0


In [25]:
before_rows = len(clean_df)

clean_df = clean_df.dropna(subset=['Date'])

removed_invalid_dates = before_rows - len(clean_df)

clean_df['Date'] = clean_df['Date'].dt.strftime('%Y-%m-%d')

add_log(
    "CR005",
    "Date column converted to YYYY-MM-DD format",
    f"{removed_invalid_dates} invalid date rows removed"
)

clean_df[['OrderID', 'Date']].head()

,OrderID,Date
0,ORD200000,2023-01-04
1,ORD200001,2024-08-23
2,ORD200002,2024-02-27
3,ORD200003,2023-10-15
4,ORD200004,2025-05-08


In [26]:
numeric_cols = ['Quantity', 'UnitPrice', 'ItemsInCart', 'TotalPrice']

for col in numeric_cols:
    clean_df[col] = pd.to_numeric(clean_df[col], errors='coerce')

numeric_missing = clean_df[numeric_cols].isnull().sum()
numeric_missing

Quantity       0
UnitPrice      0
ItemsInCart    0
TotalPrice     0
dtype: int64

In [27]:
for col in numeric_cols:
    missing_count = clean_df[col].isnull().sum()
    
    if missing_count > 0:
        median_value = clean_df[col].median()
        clean_df[col] = clean_df[col].fillna(median_value)
        
        add_log(
            f"CR_NUM_{col}",
            f"Missing values in {col} filled using median",
            f"{missing_count} missing values filled"
        )

In [28]:
text_cols = clean_df.select_dtypes(include='object').columns

for col in text_cols:
    clean_df[col] = clean_df[col].astype(str).str.strip()
    clean_df[col] = clean_df[col].str.replace(r'\s+', ' ', regex=True)

add_log(
    "CR006",
    "Text columns cleaned",
    f"{len(text_cols)} text columns cleaned by trimming spaces and removing extra whitespace"
)

In [29]:
clean_df['OrderID'] = clean_df['OrderID'].str.upper()
clean_df['CustomerID'] = clean_df['CustomerID'].str.upper()
clean_df['TrackingNumber'] = clean_df['TrackingNumber'].str.upper()
clean_df['CouponCode'] = clean_df['CouponCode'].str.upper()

clean_df['Product'] = clean_df['Product'].str.title()
clean_df['PaymentMethod'] = clean_df['PaymentMethod'].str.title()
clean_df['OrderStatus'] = clean_df['OrderStatus'].str.title()
clean_df['ReferralSource'] = clean_df['ReferralSource'].str.title()
clean_df['ShippingAddress'] = clean_df['ShippingAddress'].str.title()

add_log(
    "CR007",
    "Text formats standardized",
    "IDs and coupon codes converted to uppercase; category columns converted to title case"
)

In [30]:
expected_total = (clean_df['Quantity'] * clean_df['UnitPrice']).round(2)

total_mismatch = (clean_df['TotalPrice'].round(2) != expected_total).sum()

print("TotalPrice mismatches:", total_mismatch)

TotalPrice mismatches: 0


In [31]:
final_missing = clean_df.isnull().sum().sum()
final_full_duplicates = clean_df.duplicated().sum()
final_duplicate_order_ids = clean_df['OrderID'].duplicated().sum()
final_duplicate_tracking = clean_df['TrackingNumber'].duplicated().sum()

date_format_errors = ~clean_df['Date'].astype(str).str.match(r'^\d{4}-\d{2}-\d{2}$')
final_date_errors = date_format_errors.sum()

expected_total = (clean_df['Quantity'] * clean_df['UnitPrice']).round(2)
final_total_mismatch = (clean_df['TotalPrice'].round(2) != expected_total).sum()

print("Final Missing Values:", final_missing)
print("Final Full Duplicate Rows:", final_full_duplicates)
print("Final Duplicate OrderID:", final_duplicate_order_ids)
print("Final Duplicate TrackingNumber:", final_duplicate_tracking)
print("Final Date Format Errors:", final_date_errors)
print("Final TotalPrice Mismatches:", final_total_mismatch)
print("Final Dataset Shape:", clean_df.shape)

Final Missing Values: 0
Final Full Duplicate Rows: 0
Final Duplicate OrderID: 0
Final Duplicate TrackingNumber: 0
Final Date Format Errors: 0
Final TotalPrice Mismatches: 0
Final Dataset Shape: (1200, 14)


In [32]:
change_log_df = pd.DataFrame(change_log)
change_log_df

,Change ID,Description,Impact,Status
0,CR001,Column names cleaned by removing extra spaces,14 columns checked and standardized,Resolved
1,CR001,Column names cleaned by removing extra spaces,14 columns checked and standardized,Resolved
2,CR002,Full duplicate rows removed,0 duplicate rows removed,Resolved
3,CR004,Missing CouponCode values filled,309 missing CouponCode values replaced with NO...,Resolved
4,CR005,Date column converted to YYYY-MM-DD format,0 invalid date rows removed,Resolved
5,CR006,Text columns cleaned,10 text columns cleaned by trimming spaces and...,Resolved
6,CR007,Text formats standardized,IDs and coupon codes converted to uppercase; c...,Resolved


In [33]:
clean_df.to_excel("Cleaned_Dataset_Project1.xlsx", index=False)
change_log_df.to_excel("Change_Log_Project1.xlsx", index=False)

print("Files saved successfully!")

Files saved successfully!


Project 1: Data Cleaning & Preparation

Dataset contained 1200 rows and 14 columns.
The main data quality issue was missing CouponCode values.
309 missing CouponCode values were replaced with NO_COUPON.
Duplicate OrderID check was performed and final duplicate OrderID count is 0.
Date column was standardized into YYYY-MM-DD format.
Numeric columns were validated.
TotalPrice was verified using Quantity × UnitPrice.
Final dataset contains 0 missing values, 0 duplicate OrderIDs, and 0 date format errors.